libvast/src/option_declaration_set.cpp

```diff
@@ -0,0 +1,296 @@
+    [&](const vector&) {
+      return std::make_pair(parse_state::type_not_parsebale,
+                              default_value());
+    }), default_value_);
+  return result; 
+}
+
+const std::string&
+option_declaration_set::option_declaration::long_name() const {
+  return long_name_;
+}
+
+const std::vector<char>&
+option_declaration_set::option_declaration::short_names() const {
+  return short_names_;
+}
+
+const std::string&
+option_declaration_set::option_declaration::description() const {
+  return description_;
+}
+
+bool option_declaration_set::option_declaration::has_argument() const {
+  return has_argument_;
+}
+
+const data& option_declaration_set::option_declaration::default_value() const {
+  return default_value_;
+}
+
+option_declaration_set::option_declaration_set() {
+  add("help,h?", "print this text", false);
+}
+
+optional<const option_declaration_set::option_declaration&>
+option_declaration_set::find(std::string_view name) const {
+  // TODO: Remove explicit conversion to a string
+  if (auto it = long_opts_.find(std::string{name}); it != long_opts_.end())
+    return *it->second;
+  return {};
+}
+
+std::string option_declaration_set::usage() const {
+  //<--- argument ---> <---- desciption ---->
+  //-w [--write] arg  : path to write events to
+  auto build_argument = [](const option_declaration& x) {
+    std::stringstream arg;
+    auto& shorts = x.short_names();
+    arg << "  "; 
+    if (!shorts.empty()) {
+      arg << "-" <<shorts[0];
+      arg << " [";
+    }
+    if (shorts.size() > 1)
+      for (auto i = 1u; i < shorts.size(); ++i)
+        arg << "-" << shorts[i] << ",";
+    arg << "--" << x.long_name();
+    if (!shorts.empty())
+      arg << "]";
+    if (x.has_argument())
+      arg << " arg";
+    return arg.str();
+  };
+  // calculate the max size the argument column
+  std::vector<size_t> arg_sizes;
+  std::transform(
+    long_opts_.begin(), long_opts_.end(), std::back_inserter(arg_sizes),
+    [&](const auto& x) { return build_argument(*x.second).size(); });
+  auto help_column_width = *max_element(arg_sizes.begin(), arg_sizes.end());
+  // create usage string
+  std::stringstream res;
+  res << "Allowed options:";
+  for (auto& [_, x] : long_opts_) {
+    res << "\n"
+        << std::left << std::setw(help_column_width) << build_argument(*x)
+        << " : " << x->description();
+  }
+  return res.str();
+}
+
+size_t option_declaration_set::size() const {
+  return long_opts_.size();
+}
+
+expected<void> option_declaration_set::add(std::string_view name,
+                                           std::string_view desciption,
+                                           data default_value) {
+  // Parse short and long name.
+  std::string_view long_name;
+  std::vector<char> short_names;
+  if (auto idx = name.find(','); idx == std::string_view::npos) {
+    long_name = name;
+  } else {
+    long_name = name.substr(0, idx);
+    auto short_name = name.substr(idx + 1, name.size() - idx);
```

Comment:
You can replace these four lines with `short_names.insert(name.begin() + idx + 1, name.end())`.